src/hip/handlehip.cpp

```diff
@@ -493,8 +494,17 @@ Program Handle::LoadProgram(const std::string& program_name,
 
     std::string orig_params = params; // make a copy for target ID fallback
 
-    if(!miopen::EndsWith(program_name, ".mlir"))
-        params = params + " -mcpu=" + this->GetTargetProperties().Name();
+    if(miopen::EndsWith(program_name, ".mlir"))
+    { // no -mcpu
+    }
+    else if(miopen::EndsWith(program_name, ".s"))
+    {
+        params += " -mcpu=" + LcOptionTargetStrings{this->GetTargetProperties()}.targetId;
```

Comment:
@atamazov  Why is the target ID required here ? This would break kdb since our kdb is target ID agnostic ?

Reply:
@JehandadKhan 
> Why is the target ID required here ?

It's assembler requirement, it wants to know XNACK state etc. And it doesn't accept options like `-mno-xnack` anymore. You can find a link to the documentation below:

> Offline (clang) amdgcn assembly:
> 
> * Removed "-mno-xnack".
> * Added features to "-mcpu" value (e.g. `"-mcpu=gfx908"` -> `"-mcpu=gfx908:xnack-"`,  see [here](https://llvm.org/docs/AMDGPUUsage.html#amdgpu-target-features) for more info)

Reply:
@JehandadKhan 
> This would break kdb since our kdb is target ID agnostic ?

No, I think. IIRC this information is included in the "options" string that is used in the KDB key. Therefore KDB will hold a separate kernel binary for each targetId.

Reply:
@atamazov @JehandadKhan we do see that naive kernels in cache are broken after this PR, and removing KDB helps to make the tests pass
```
Observations:
:/opt/rocm# ./bin/MIOpenDriver conv -n 128 -c 3 -H 224 -W 224 -k 64 -y 7 -x 7 -p 3 -q 3 -u 2 -v 2 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpenDriver conv -n 128 -c 3 -H 224 -W 224 -k 64 -y 7 -x 7 -p 3 -q 3 -u 2 -v 2 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
PRNG seed: 12345678
GPU core dump failed
:0:rocdevice.cpp            :2958: 9469160433 us: [pid:11396 tid:0x7f91349ff640] Callback: Queue 0x7f8931600000 aborting with error : HSA_STATUS_ERROR_MEMORY_APERTURE_VIOLATION: The agent attempted to access memory beyond the largest legal address. code: 0x29
Aborted (core dumped)

Observation when we remove the KDB file:
:/opt/rocm# mv /opt/rocm/share/miopen/db/gfx942.kdb /
:/opt/rocm# ./bin/MIOpenDriver conv -n 128 -c 3 -H 224 -W 224 -k 64 -y 7 -x 7 -p 3 -q 3 -u 2 -v 2 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
MIOpenDriver conv -n 128 -c 3 -H 224 -W 224 -k 64 -y 7 -x 7 -p 3 -q 3 -u 2 -v 2 -l 1 -j 1 -m conv -g 1 -F 1 -t 1
PRNG seed: 12345678
MIOpen Forward Conv. Algorithm: 3, Solution: 37/ConvBinWinogradRxSf3x2
GPU Kernel Time Forward Conv. Elapsed: 0.864864 ms (average)
stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
stats: fwd-conv7x7u2, 128, 3, 112, 112, 7, 7, 64,  30211571712, 77107968, 411041792, 34932, 564, 0.864864
Forward Convolution Verifies OK on GPU reference (3.47741e-08 < 1.5e-06)
```

Reply:
It's true that the current kdb does not withstand this change. I'll open a PR shortly correcting db_sync for this change along with mi100, mi200, and mi300 kdb files.
Another concern might be that compilations from our tuning machines may not translate to the user machine if they have certain features enabled/disabled.

Reply:
@JehandadKhan @junliume @cderb 
> our kdb is target ID agnostic.

I am afraid this should not be so, see https://github.com/ROCm/MIOpen/pull/2309#issuecomment-1675431339

When we do not specify the exact value of the target feature, then the HIP/OCL compiler selects the safest variant of code generation. Which means that the kernels written in HIP or OpenCL may yield worse performance than potentially possible when:
- If a GPU does not have XNACK enabled.
- If a GPU does not have SRAMECC enabled.

Our assembly kernels are written using CO v3 and therefore if target features are unset (during compilation), then such kernels will correctly work **_only_** on `xnack+:sramecc+` targets. More info at https://llvm.org/docs/AMDGPUUsage.html#amdgpu-target-features-table

A while ago I've asked about some strange change at https://github.com/ROCm/MIOpen/pull/2157#discussion_r1294646921 (but didn't received the answer).

There is unfinished discussion about all this stuff, https://github.com/ROCm/MIOpen/pull/674#discussion_r558551331.

Reply:
I think this answered largely my question at https://github.com/ROCm/MIOpen/pull/2996#issuecomment-2135698107

Reply:
@JehandadKhan @junliume @cderb FYI https://github.com/ROCm/MIOpen/pull/2891#discussion_r1616343986 is just updated (clarified), please have a look.